Add a TLSv1.3 test that ensures a zero content type is handled correctly

4a6f656c opened the pull request:
### Description
<!-- Describe your changes in detail below -->

In TLSv1.3, once record protection has been engaged, all records are sent
as application data records on the wire, with the real content type being
added to the end of the plaintext record, followed by optional zero-byte
padding. If the content type is zero (the plaintext consists only of bytes
with value zero) then an unexpected message alert must be sent.

### Motivation and Context
<!-- Describe why the change is introduced, if it solves an issue add "fixes #1"
with a correct number -->

Updates #355

### Checklist
<!-- go over following points. check them with an `x` if they do apply,
(they turn into clickable checkboxes once the PR is submitted, so no need
to do everything at once)

if you're unsure about any of those items, just ask in comment to PR

if the PR resolves an issue, please add further checkboxes that describe the
action items or test scenarios from it
-->

- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [ ] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [ ] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [ ] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [ ] NSS
  - [ ] GnuTLS
- [ ] required version of tlslite-ng updated in requirements.txt and README.md

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)
<!-- Reviewable:end -->

tomato42 reviewed (changes requested):
sorry for the delay, I was on holidays

Thanks for the PR! Looks good overall, but I have few nits for it.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-:-MFVvJAM-rYXRzV1EhNv:b-inye9f)* status: all files reviewed, 6 unresolved discussions (waiting on @4a6f656c)

*[a discussion](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-lsj2zc4PsUrIYNY:-MFW-lsj2zc4PsUrIYNZ:b-izlrfm) (no related file):*
Technically speaking, this script does not fix #355:
1. it does not send empty (zero-length plaintext) messages (if only content_type is sent then the payload is 1 byte long)
2. it does not send messages with content_type of 0 using early data (0-RTT) keys

The latter is blocked by #205, so we may split it to a new issue, but former should be doable. If you don't want to fix them, you can just remove the "fixes #355" from the PR description and commit message.

---
___
*[scripts/test-tls13-zero-content-type.py, line 34 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVvBaiEjI4MohymlRf:-MFVvBaiEjI4MohymlRg:b35xg3h) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L34)):*
> ```Python
> 
> 
> version = 3
> ```

you were developing it privately before?
___
*[scripts/test-tls13-zero-content-type.py, line 111 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-_Ix8bQrJPddn74d:-MFW-_Ix8bQrJPddn74e:b-8fn9ar) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L111)):*
> ```Python
>         .create(groups)
>     sig_algs = [SignatureScheme.rsa_pss_rsae_sha256,
>                 SignatureScheme.rsa_pss_pss_sha256]
> ```

why no support for ECDSA certificates?
___
*[scripts/test-tls13-zero-content-type.py, line 115 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVyB7wBDitb5o6hLb_:-MFVyB7wBDitb5o6hLba:b-twnbvx) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L115)):*
> ```Python
>         .create(sig_algs)
>     ext[ExtensionType.signature_algorithms_cert] = SignatureAlgorithmsCertExtension()\
>         .create(RSA_SIG_ALL)
> ```

why limit to just RSA signatures?
___
*[scripts/test-tls13-zero-content-type.py, line 137 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW1IAD5KtUe93f_kaI:-MFW1IAD5KtUe93f_kaJ:b-vdlf1y) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L137)):*
> ```Python
> 
>     node = node.add_child(ExpectAlert())
>     node.next_sibling = ExpectClose()
> ```

`node.add_child(ExpectClose())` missing, see https://tlsfuzzer.readthedocs.io/en/latest/writing-tests.html#closing-the-connection-alternatives-in-decision-graphs
___
*[scripts/test-tls13-zero-content-type.py, line 442 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-Dqk-wVVFwL1Nofs:-MFW-DqlEAAt1w9HcTVk:b-qddk7y) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L442)):*
> ```Python
>     print("Check if handling of records with an internal content type of zero is ")
>     print("correct.\n")
>     print("version: {0}\n".format(version))
> ```

nit: in `test-tls13-conversation.py` the version is printed after "Test end"


<!-- Sent from Reviewable.io -->

4a6f656c commented:
FWIW I tried to do this on reviewable, however it requires permissions that I cannot grant it.

> sorry for the delay, I was on holidays

No problem, thanks for the review.

> Thanks for the PR! Looks good overall, but I have few nits for it.
> 
> > Reviewed 3 of 3 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-:-MFVvJAM-rYXRzV1EhNv:b-inye9f)_ status: all files reviewed, 6 unresolved discussions (waiting on @4a6f656c)
> 
> _[a discussion](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-lsj2zc4PsUrIYNY:-MFW-lsj2zc4PsUrIYNZ:b-izlrfm) (no related file):_
> Technically speaking, this script does not fix #355:
> 
> 1. it does not send empty (zero-length plaintext) messages (if only content_type is sent then the payload is 1 byte long)
> 2. it does not send messages with content_type of 0 using early data (0-RTT) keys
> 
> The latter is blocked by #205, so we may split it to a new issue, but former should be doable. If you don't want to fix them, you can just remove the "fixes #355" from the PR description and commit message.

 Ack. I could not see an easy way to do (1) as the RawMessageGenerator() requires a content_type and passes this through to tlslite.messages.Message - sendRecord then implicitly appends the contentType byte for TLSv1.3, hence this appears to require changes to tlslite to be able to implement. If there is something I'm missing please let me know.

I'll remove the fixes - do you want an "updates" or similar for this issue?

> _[scripts/test-tls13-zero-content-type.py, line 34 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVvBaiEjI4MohymlRf:-MFVvBaiEjI4MohymlRg:b35xg3h) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L34)):_
> 
> > ```python
> > version = 3
> > ```
> 
> you were developing it privately before?

This was derived from `test-tls13-zero-length-data.py` and retained - would you prefer this be reset to 1?

> _[scripts/test-tls13-zero-content-type.py, line 111 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-_Ix8bQrJPddn74d:-MFW-_Ix8bQrJPddn74e:b-8fn9ar) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L111)):_
> 
> > ```python
> >         .create(groups)
> >     sig_algs = [SignatureScheme.rsa_pss_rsae_sha256,
> >                 SignatureScheme.rsa_pss_pss_sha256]
> > ```
> 
> why no support for ECDSA certificates?

This is what is used in `test-tls13-zero-length-data.py` - can change if preferred.

> _[scripts/test-tls13-zero-content-type.py, line 115 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVyB7wBDitb5o6hLb_:-MFVyB7wBDitb5o6hLba:b-twnbvx) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L115)):_
> 
> > ```python
> >         .create(sig_algs)
> >     ext[ExtensionType.signature_algorithms_cert] = SignatureAlgorithmsCertExtension()\
> >         .create(RSA_SIG_ALL)
> > ```
> 
> why limit to just RSA signatures?

This is what is done in `test-tls13-zero-length-data.py` - can change if preferred.

> _[scripts/test-tls13-zero-content-type.py, line 137 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW1IAD5KtUe93f_kaI:-MFW1IAD5KtUe93f_kaJ:b-vdlf1y) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L137)):_
> 
> > ```python
> >     node = node.add_child(ExpectAlert())
> >     node.next_sibling = ExpectClose()
> > ```
> 
> `node.add_child(ExpectClose())` missing, see https://tlsfuzzer.readthedocs.io/en/latest/writing-tests.html#closing-the-connection-alternatives-in-decision-graphs

Fixed (also missing from `test-tls13-zero-length-data.py` - can send a follow up)

> _[scripts/test-tls13-zero-content-type.py, line 442 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-Dqk-wVVFwL1Nofs:-MFW-DqlEAAt1w9HcTVk:b-qddk7y) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L442)):_
> 
> > ```python
> >     print("Check if handling of records with an internal content type of zero is ")
> >     print("correct.\n")
> >     print("version: {0}\n".format(version))
> > ```
> 
> nit: in `test-tls13-conversation.py` the version is printed after "Test end"

Fixed (fwiw test-tls13-zero-length-data.py does it differently :)

tomato42 commented:
> > Thanks for the PR! Looks good overall, but I have few nits for it.
> > > Reviewed 3 of 3 files at r1.
> > > _[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-:-MFVvJAM-rYXRzV1EhNv:b-inye9f)_ status: all files reviewed, 6 unresolved discussions (waiting on @4a6f656c)
> > 
> > 
> > _[a discussion](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-lsj2zc4PsUrIYNY:-MFW-lsj2zc4PsUrIYNZ:b-izlrfm) (no related file):_
> > Technically speaking, this script does not fix #355:
> > 
> > 1. it does not send empty (zero-length plaintext) messages (if only content_type is sent then the payload is 1 byte long)
> > 2. it does not send messages with content_type of 0 using early data (0-RTT) keys
> > 
> > The latter is blocked by #205, so we may split it to a new issue, but former should be doable. If you don't want to fix them, you can just remove the "fixes #355" from the PR description and commit message.
> 
> Ack. I could not see an easy way to do (1) as the RawMessageGenerator() requires a content_type and passes this through to tlslite.messages.Message - sendRecord then implicitly appends the contentType byte for TLSv1.3, hence this appears to require changes to tlslite to be able to implement. If there is something I'm missing please let me know.

ah, true, I thought that it would be possible with one of the wrappers, like the `replace_plaintext()`, but it isn't updated to work with AEAD ciphers; we'd need a similar wrapper for the `_encryptThenSeal` method of RecordLayer

> I'll remove the fixes - do you want an "updates" or similar for this issue?

`updates #xx` is fine
 
> > _[scripts/test-tls13-zero-content-type.py, line 34 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVvBaiEjI4MohymlRf:-MFVvBaiEjI4MohymlRg:b35xg3h) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L34)):_
> > > ```python
> > > version = 3
> > > ```
> > 
> > 
> > you were developing it privately before?
> 
> This was derived from `test-tls13-zero-length-data.py` and retained - would you prefer this be reset to 1?

yes, since it's a new script it should start with 1

> > _[scripts/test-tls13-zero-content-type.py, line 111 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-_Ix8bQrJPddn74d:-MFW-_Ix8bQrJPddn74e:b-8fn9ar) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L111)):_
> > > ```python
> > >         .create(groups)
> > >     sig_algs = [SignatureScheme.rsa_pss_rsae_sha256,
> > >                 SignatureScheme.rsa_pss_pss_sha256]
> > > ```
> > 
> > 
> > why no support for ECDSA certificates?
> 
> This is what is used in `test-tls13-zero-length-data.py` - can change if preferred.

please do, in general it's the `test-tls13-conversation.py` and `test-conversation.py` that are the templates for new scripts
 
> > _[scripts/test-tls13-zero-content-type.py, line 115 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVyB7wBDitb5o6hLb_:-MFVyB7wBDitb5o6hLba:b-twnbvx) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L115)):_
> > > ```python
> > >         .create(sig_algs)
> > >     ext[ExtensionType.signature_algorithms_cert] = SignatureAlgorithmsCertExtension()\
> > >         .create(RSA_SIG_ALL)
> > > ```
> > 
> > 
> > why limit to just RSA signatures?
> 
> This is what is done in `test-tls13-zero-length-data.py` - can change if preferred.

please do
 
> > _[scripts/test-tls13-zero-content-type.py, line 137 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW1IAD5KtUe93f_kaI:-MFW1IAD5KtUe93f_kaJ:b-vdlf1y) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L137)):_
> > > ```python
> > >     node = node.add_child(ExpectAlert())
> > >     node.next_sibling = ExpectClose()
> > > ```
> > 
> > 
> > `node.add_child(ExpectClose())` missing, see https://tlsfuzzer.readthedocs.io/en/latest/writing-tests.html#closing-the-connection-alternatives-in-decision-graphs
> 
> Fixed (also missing from `test-tls13-zero-length-data.py` - can send a follow up)

that would be appreciated

while we try to keep all scripts up to date with newest conventions, it's not always done, so some minor problems like this do linger
 
> > _[scripts/test-tls13-zero-content-type.py, line 442 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-Dqk-wVVFwL1Nofs:-MFW-DqlEAAt1w9HcTVk:b-qddk7y) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L442)):_
> > > ```python
> > >     print("Check if handling of records with an internal content type of zero is ")
> > >     print("correct.\n")
> > >     print("version: {0}\n".format(version))
> > > ```
> > 
> > 
> > nit: in `test-tls13-conversation.py` the version is printed after "Test end"
> 
> Fixed (fwiw test-tls13-zero-length-data.py does it differently :)

tomato42 reviewed (commented):
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)* status: all files reviewed, 4 unresolved discussions (waiting on @4a6f656c)



<!-- Sent from Reviewable.io -->

4a6f656c commented:
PTAL. 

> > I'll remove the fixes - do you want an "updates" or similar for this issue?
> 
> `updates #xx` is fine

Done.
 
> > > _[scripts/test-tls13-zero-content-type.py, line 34 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVvBaiEjI4MohymlRf:-MFVvBaiEjI4MohymlRg:b35xg3h) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L34)):_
> > > > ```python
> > > > version = 3
> > > > ```
> > > 
> > > 
> > > you were developing it privately before?
> > 
> > 
> > This was derived from `test-tls13-zero-length-data.py` and retained - would you prefer this be reset to 1?
> 
> yes, since it's a new script it should start with 1

Done.
 
> > > _[scripts/test-tls13-zero-content-type.py, line 111 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFW-_Ix8bQrJPddn74d:-MFW-_Ix8bQrJPddn74e:b-8fn9ar) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L111)):_
> > > > ```python
> > > >         .create(groups)
> > > >     sig_algs = [SignatureScheme.rsa_pss_rsae_sha256,
> > > >                 SignatureScheme.rsa_pss_pss_sha256]
> > > > ```
> > > 
> > > 
> > > why no support for ECDSA certificates?
> > 
> > 
> > This is what is used in `test-tls13-zero-length-data.py` - can change if preferred.
> 
> please do, in general it's the `test-tls13-conversation.py` and `test-conversation.py` that are the templates for new scripts

Fixed.
 
> > > _[scripts/test-tls13-zero-content-type.py, line 115 at r1](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFVyB7wBDitb5o6hLb_:-MFVyB7wBDitb5o6hLba:b-twnbvx) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/a69342d22c3ea49294c8d3bdfede9a8849ebf6a2/scripts/test-tls13-zero-content-type.py#L115)):_
> > > > ```python
> > > >         .create(sig_algs)
> > > >     ext[ExtensionType.signature_algorithms_cert] = SignatureAlgorithmsCertExtension()\
> > > >         .create(RSA_SIG_ALL)
> > > > ```
> > > 
> > > 
> > > why limit to just RSA signatures?
> > 
> > 
> > This is what is done in `test-tls13-zero-length-data.py` - can change if preferred.
> 
> please do


Fixed.

tomato42 reviewed (approved):
> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

tomato42 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)* status: all files reviewed, 2 unresolved discussions (waiting on @4a6f656c)

___
*[scripts/test-tls13-zero-content-type.py, line 259 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfwCu_9WxqD_s1Vpg9:-MFfwCu_9WxqD_s1VpgA:bfw00ce) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L259)):*
> ```Python
>     node = node.add_child(RawMessageGenerator(content_type=0, data=bytearray(0)))
> 
>     node = node.add_child(ExpectAlert(AlertLevel.fatal,
> ```

I've run the test against GnuTLS and noticed that actually the server can send a NewSessionTicket here, before an Alert: if the connection doesn't use client certificate based authentication the server can send the NewSessionTicket without waiting for Finished from client
___
*[scripts/test-tls13-zero-content-type.py, line 296 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfzQFg5XFYDCoajlyF:-MFfzQFh5K6PAyNYRDke:b-2guby3) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L296)):*
> ```Python
>     node = node.add_child(RawMessageGenerator(content_type=0, data=bytearray(42)))
> 
>     node = node.add_child(ExpectAlert(AlertLevel.fatal,
> ```

same here


<!-- Sent from Reviewable.io -->

4a6f656c commented:
> > _[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)_ status: all files reviewed, 2 unresolved discussions (waiting on @4a6f656c)
> 
> _[scripts/test-tls13-zero-content-type.py, line 259 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfwCu_9WxqD_s1Vpg9:-MFfwCu_9WxqD_s1VpgA:bfw00ce) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L259)):_
> 
> > ```python
> >     node = node.add_child(RawMessageGenerator(content_type=0, data=bytearray(0)))
> > 
> >     node = node.add_child(ExpectAlert(AlertLevel.fatal,
> > ```
> 
> I've run the test against GnuTLS and noticed that actually the server can send a NewSessionTicket here, before an Alert: if the connection doesn't use client certificate based authentication the server can send the NewSessionTicket without waiting for Finished from client

Good catch, thanks - fixed.

> _[scripts/test-tls13-zero-content-type.py, line 296 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfzQFg5XFYDCoajlyF:-MFfzQFh5K6PAyNYRDke:b-2guby3) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L296)):_
> 
> > ```python
> >     node = node.add_child(RawMessageGenerator(content_type=0, data=bytearray(42)))
> > 
> >     node = node.add_child(ExpectAlert(AlertLevel.fatal,
> > ```
> 
> same here

Fixed.

tomato42 reviewed (changes requested):
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)* status: all files reviewed, 2 unresolved discussions (waiting on @4a6f656c)

___
*[scripts/test-tls13-zero-content-type.py, line 259 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfwCu_9WxqD_s1Vpg9:-MFjYykc489jOQPfHgOx:b-r9jp3) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L259)):*
<details><summary><i>Previously, 4a6f656c (Joel Sing) wrote…</i></summary><blockquote>

Good catch, thanks - fixed. 
</blockquote></details>

*can* send, not *must* send, so ExpectAlert needs to be a sibling, like in other tests
___
*[scripts/test-tls13-zero-content-type.py, line 296 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfzQFg5XFYDCoajlyF:-MFjZEPp3zOJz_32hEF9:b-2guby3) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L296)):*
<details><summary><i>Previously, 4a6f656c (Joel Sing) wrote…</i></summary><blockquote>

Fixed. 
</blockquote></details>

same here


<!-- Sent from Reviewable.io -->

4a6f656c commented:
> > Reviewed 1 of 1 files at r7.
> > _[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694)_ status: all files reviewed, 2 unresolved discussions (waiting on @4a6f656c)
> 
> _[scripts/test-tls13-zero-content-type.py, line 259 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfwCu_9WxqD_s1Vpg9:-MFjYykc489jOQPfHgOx:b-r9jp3) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L259)):_
> 
> Previously, 4a6f656c (Joel Sing) wrote…
> _can_ send, not _must_ send, so ExpectAlert needs to be a sibling, like in other tests

Ugh, thanks.
 
> _[scripts/test-tls13-zero-content-type.py, line 296 at r6](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-MFfzQFg5XFYDCoajlyF:-MFjZEPp3zOJz_32hEF9:b-2guby3) ([raw file](https://github.com/tomato42/tlsfuzzer/blob/d7b45cfa26193f55002da8f42a53a15e82cfda06/scripts/test-tls13-zero-content-type.py#L296)):_
> 
> Previously, 4a6f656c (Joel Sing) wrote…
> same here

Done.

tomato42 reviewed (approved):
Sorry, I hoped that I won't need to force you to do that, but please squash the changes and rebase, not merge master to this branch: the the last merge of master to this branch caused tests in Travis to fail. 

This is because we have Travis set up in a way it runs tests on each commit in turn, so if there are 3 commits it runs the test suite 3 times (to make sure that the test suite is always runnable, in case it is needed in a future bisect), but if there are merge commits it will rerun test suite on all commits until the first common commit of master an this branch, irrespective of merges.

> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-:-MFox_2P1avbcuSWJ1bb:be1rxhn)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

4a6f656c commented:
> Sorry, I hoped that I won't need to force you to do that, but please squash the changes and rebase, not merge master to this branch: the the last merge of master to this branch caused tests in Travis to fail.

Nothing to force, only need to ask :)
 
> This is because we have Travis set up in a way it runs tests on each commit in turn, so if there are 3 commits it runs the test suite 3 times (to make sure that the test suite is always runnable, in case it is needed in a future bisect), but if there are merge commits it will rerun test suite on all commits until the first common commit of master an this branch, irrespective of merges.

No problem - different projects have different preferences. Have squashed and rebased.
 
> > Reviewed 1 of 1 files at r8.
> > _[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-:-MFox_2P1avbcuSWJ1bb:be1rxhn)_ status: :shipit: complete! all files reviewed, all discussions resolved

tomato42 commented:
> > Sorry, I hoped that I won't need to force you to do that, but please squash the changes and rebase, not merge master to this branch: the the last merge of master to this branch caused tests in Travis to fail.
> 
> Nothing to force, only need to ask :)

Yeah, but I've already had asked for much so I didn't want to strain your patience :)

do you want to file bugs against gnutls and NSS that they handle those records incorrectly, or should I do it?
 
> > This is because we have Travis set up in a way it runs tests on each commit in turn, so if there are 3 commits it runs the test suite 3 times (to make sure that the test suite is always runnable, in case it is needed in a future bisect), but if there are merge commits it will rerun test suite on all commits until the first common commit of master an this branch, irrespective of merges.
> 
> No problem - different projects have different preferences. Have squashed and rebased.

perfect, thank you!
 
> > > Reviewed 1 of 1 files at r8.
> > > _[Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/694#-:-MFox_2P1avbcuSWJ1bb:be1rxhn)_ status: :shipit: complete! all files reviewed, all discussions resolved

4a6f656c commented:
> do you want to file bugs against gnutls and NSS that they handle those records incorrectly, or should I do it?

Please go ahead - if you would not mind providing links to the reports, that would be great.

tomato42 commented:
> > do you want to file bugs against gnutls and NSS that they handle those records incorrectly, or should I do it?
> 
> Please go ahead - if you would not mind providing links to the reports, that would be great.

added them inline into the opening comment

thanks again for the contribution!
